pkg/loadbalancer/experimental/tables.go

```diff
@@ -0,0 +1,299 @@
+	// This is managed by [Services].
+	ReferencedBy container.ImmSet[loadbalancer.ServiceName]
+}
+
+func (be *Backend) String() string {
+	return strings.Join(be.TableRow(), " ")
+}
+
+func (be *Backend) TableHeader() []string {
+	return []string{
+		"Address",
+		"State",
+		"Source",
+		"ReferencedBy",
+	}
+}
+
+func toStrings[T fmt.Stringer](xs []T) []string {
+	out := make([]string, len(xs))
+	for i := range xs {
+		out[i] = xs[i].String()
+	}
+	return out
+}
+
+func (be *Backend) TableRow() []string {
+	state, err := be.State.String()
+	if err != nil {
+		state = err.Error()
+	}
+	return []string{
+		be.L3n4Addr.StringWithProtocol(),
+		state,
+		string(be.Source),
+		strings.Join(toStrings(be.ReferencedBy.AsSlice()), ", "),
+	}
+}
+
+func (be *Backend) InLocalCluster() bool {
+	return be.Cluster == ""
+}
+
+func (be *Backend) InRemoteCluster() bool {
+	return be.Cluster != ""
+}
+
+func (be *Backend) removeRef(name loadbalancer.ServiceName) (*Backend, bool) {
+	beCopy := *be
+	beCopy.ReferencedBy = beCopy.ReferencedBy.Delete(name)
+	return &beCopy, beCopy.ReferencedBy.Len() == 0
+}
+
+// Clone returns a shallow clone of the backend.
+func (be *Backend) Clone() *Backend {
+	be2 := *be
+	return &be2
+}
+
+var (
+	BackendAddrIndex = statedb.Index[*Backend, loadbalancer.L3n4Addr]{
+		Name: "addr",
+		FromObject: func(obj *Backend) index.KeySet {
+			return index.NewKeySet(l3n4AddrKey(obj.L3n4Addr))
+		},
+		FromKey: l3n4AddrKey,
+		Unique:  true,
+	}
+
+	BackendServiceIndex = statedb.Index[*Backend, loadbalancer.ServiceName]{
+		Name: "service-name",
+		FromObject: func(obj *Backend) index.KeySet {
+			return index.StringerSlice(obj.ReferencedBy.AsSlice())
+		},
+		FromKey: index.Stringer[loadbalancer.ServiceName],
+		Unique:  false,
+	}
+)
+
+func NewBackendsTable(db *statedb.DB) (statedb.RWTable[*Backend], error) {
+	tbl, err := statedb.NewTable(
+		BackendTableName,
+		BackendAddrIndex,
+		BackendServiceIndex,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return tbl, db.RegisterTable(tbl)
+}
+
+func GetBackendsForFrontend(txn statedb.ReadTxn, tbl statedb.Table[*Backend], svc *Frontend) statedb.Iterator[*Backend] {
+	setPreferred := func(beImmutable *Backend) *Backend {
+		be := beImmutable.Clone()
+
+		// First set preferred based on service affinity
+		switch svc.ServiceAffinity {
+		case ServiceAffinityNone:
+			fallthrough
+		case ServiceAffinityLocal:
+			be.Preferred = loadbalancer.Preferred(be.InLocalCluster())
+		case ServiceAffinityRemote:
+			be.Preferred = loadbalancer.Preferred(be.InRemoteCluster())
+		}
+
+		return be
+	}
+
+	iter, _ := tbl.Get(txn, BackendServiceIndex.Query(svc.Name))
+	if !svc.IncludeRemoteBackends {
+		iter = statedb.Filter(iter, (*Backend).InLocalCluster)
+	}
+	return statedb.Map(iter, setPreferred)
+}
+
+type FrontendParams struct {
+	Name     loadbalancer.ServiceName // Fully qualified service name
+	Source   source.Source
+	Frontend loadbalancer.L3n4Addr // Frontend address
```

Comment:
Struct `Frontend` has a field of type `FrontendParams`, which in turn has a field called `Frontend` again yet of a different type. I'd suggest renaming the last one to something like `Address`.

Reply:
Ah good point, will rename.